  ---
title: "RNA-seq Differentially Expressed Genes"
author: "Gautier Richard"
date: "10/21/2020"
output: html_document
---

# RNA-seq Downstream Analyses


The goal of this document is to take as input the RNA-seq datasets produced by SnakePipes from the FeatureCounts step to calculate DEGs and all the subsequent downstream analyses.

It includes notably the computation of normalised expression at the gene by gene and homeoglobal scales. It also computes the Additivity and transgressive expression of genes between the hybrid and its two parents. 

It also checks the expression on the A vs C subgenomes of brassica.

# Setup

## Dependencies loading

```{r, setup}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(vroom)
library(limma) 
library(edgeR)
library(splitstackshape)
library(RColorBrewer)
library(kableExtra)
library(ggplot2)
library(ggrastr)
library(tidyr)
library(cowplot)
```


## Function to calculate parental average while normalising the expression of genes

```{r, parental mean}
voom_lcpm = function (counts, design = NULL, lib.size = NULL, normalize.method = "none", 
    block = NULL, correlation = NULL, weights = NULL, span = 0.5, 
    plot = FALSE, save.plot = FALSE) 
{
    out <- list()
    if (is(counts, "DGEList")) {
        out$genes <- counts$genes
        out$targets <- counts$samples
        if (is.null(design) && diff(range(as.numeric(counts$sample$group))) > 
            0) 
            design <- model.matrix(~group, data = counts$samples)
        if (is.null(lib.size)) 
            lib.size <- counts$samples$lib.size * counts$samples$norm.factors
        counts <- counts$counts
    }
    else {
        isExpressionSet <- suppressPackageStartupMessages(is(counts, 
            "ExpressionSet"))
        if (isExpressionSet) {
            if (length(Biobase::fData(counts))) 
                out$genes <- Biobase::fData(counts)
            if (length(Biobase::pData(counts))) 
                out$targets <- Biobase::pData(counts)
            counts <- Biobase::exprs(counts)
        }
        else {
            counts <- as.matrix(counts)
        }
    }
    n <- nrow(counts)
    if (n < 2L) 
        stop("Need at least two genes to fit a mean-variance trend")
    if (is.null(design)) {
        design <- matrix(1, ncol(counts), 1)
        rownames(design) <- colnames(counts)
        colnames(design) <- "GrandMean"
    }
    if (is.null(lib.size)) 
        lib.size <- colSums(counts)
#    y <- t(log2(t(counts + 0.5)/(lib.size + 1) * 1e+06))
    y <- counts
    y <- normalizeBetweenArrays(y, method = normalize.method)
    fit <- lmFit(y, design, block = block, correlation = correlation, 
        weights = weights)
    if (is.null(fit$Amean)) 
        fit$Amean <- rowMeans(y, na.rm = TRUE)
    NWithReps <- sum(fit$df.residual > 0L)
    if (NWithReps < 2L) {
        if (NWithReps == 0L) 
            warning("The experimental design has no replication. Setting weights to 1.")
        if (NWithReps == 1L) 
            warning("Only one gene with any replication. Setting weights to 1.")
        out$E <- y
        out$weights <- y
        out$weights[] <- 1
        out$design <- design
        if (is.null(out$targets)) 
            out$targets <- data.frame(lib.size = lib.size)
        else out$targets$lib.size <- lib.size
        return(new("EList", out))
    }
    sx <- fit$Amean + mean(log2(lib.size + 1)) - log2(1e+06)
    sy <- sqrt(fit$sigma)
    allzero <- rowSums(counts) == 0
    if (any(allzero)) {
        sx <- sx[!allzero]
        sy <- sy[!allzero]
    }
    l <- lowess(sx, sy, f = span)
    if (plot) {
        plot(sx, sy, xlab = "log2( count size + 0.5 )", ylab = "Sqrt( standard deviation )", 
            pch = 16, cex = 0.25)
        title("voom: Mean-variance trend")
        lines(l, col = "red")
    }
    f <- approxfun(l, rule = 2, ties = list("ordered", mean))
    if (fit$rank < ncol(design)) {
        j <- fit$pivot[1:fit$rank]
        fitted.values <- fit$coef[, j, drop = FALSE] %*% t(fit$design[, 
            j, drop = FALSE])
    }
    else {
        fitted.values <- fit$coef %*% t(fit$design)
    }
    fitted.cpm <- 2^fitted.values
    fitted.count <- 1e-06 * t(t(fitted.cpm) * (lib.size + 1))
    fitted.logcount <- log2(fitted.count)
    w <- 1/f(fitted.logcount)^4
    dim(w) <- dim(fitted.logcount)
    out$E <- y
    out$weights <- w
    out$design <- design
    if (is.null(out$targets)) 
        out$targets <- data.frame(lib.size = lib.size)
    else out$targets$lib.size <- lib.size
    if (save.plot) {
        out$voom.xy <- list(x = sx, y = sy, xlab = "log2( count size + 0.5 )", 
            ylab = "Sqrt( standard deviation )")
        out$voom.line <- l
    }
    new("EList", out)
}
```

## DEG calculation function

```{r, DEG calculation}
getDEG <- function(path, files, group, columns, contr.matrix, lfc, genome, normalisation, col, keep_homeo){
  
# LOAD DATA

data_list = list()
count = 0
for (f in files){
  count = count + 1 
  file_full =  paste0(path, files[[count]])
   data_list[[f]] = suppressMessages( vroom(file_full, skip = 1, col_names = T) )
}

genes_expression =  do.call(cbind,data_list)
to_keep = seq(from = 7, to = ncol(genes_expression), by = 7)
genes_expression = genes_expression[,c(1:6,to_keep)]

colnames(genes_expression) = columns
genes_expression_basic = genes_expression
rownames(genes_expression_basic) = genes_expression_basic$gene


# FILTER GENES BY HOMEOLOGUES

if (keep_homeo == T){
homeo = vroom('/mnt/genouest/groups/bipaa/work/Acyrthosyphon_pisum/EGI/grichard/BP/STIRRER/Resource/Homeologie_Bna_DarV10_Meth.txt', col_names = F)
homeoA = as.data.frame(homeo$X1)
homeoC = as.data.frame(homeo$X2)
colnames(homeoA) = "gene"
colnames(homeoC) = "gene"
homeo = as.data.frame(rbind(homeoA, homeoC))
homeo = as.data.frame(homeo[order(homeo),])
colnames(homeo) = "gene"
homeo = unique(homeo)
rownames(homeo) = homeo$gene

genes_expression_basic = merge(genes_expression_basic, homeo, by = "row.names" )
genes_expression_basic = genes_expression_basic[,3:ncol(genes_expression_basic)-1]
colnames(genes_expression_basic)[1] = "gene"
rownames(genes_expression_basic) = genes_expression_basic$gene

}


# GENES FILTRATION DEPENDING ON CHOSEN GENOME

genes_expression_basic = genes_expression_basic[grep(genes_expression_basic$chr, pattern = genome),]

# LOAD GENES ANNOTATION FROM FEATURECOUNTS OUTPUT

genes             = genes_expression_basic[,c(1:6)]
genes[,2]         = sub(x = genes[,2],pattern = "^.*;",replacement = "")
genes[,3]         = sub(x = genes[,3],pattern = ";.*$",replacement = "")
genes[,4]         = sub(x = genes[,4],pattern = "^.*;",replacement = "")
genes[,5]         = sub(x = genes[,5],pattern = "^.*;",replacement = "")


# CREATE DGE OBJECT

counts            = genes_expression_basic[,c(7:ncol(genes_expression_basic))]
row.names(counts) <- genes_expression_basic[,1]
lib.size          = apply( genes_expression_basic[,7:ncol(genes_expression_basic)], 2, sum )
norm.factors      = rep(1, length(lib.size))
samplenames      = colnames(counts)
samples           = as.data.frame( cbind(files, group) )
rownames(samples) = samplenames

x = DGEList(counts = counts,
             lib.size = lib.size,
             genes = genes,
             samples = samples, 
             remove.zeros = F)

cat(cbind(samplenames,lib.size))

rownames(x$counts) = x$genes$gene


# KEEP ONLY EXPRESSED GENES AND PLOT

keep.exprs <- filterByExpr(x, group=group)
x_expr <- x[keep.exprs,, keep.lib.sizes=FALSE]

nsamples <- ncol(x)

L <- mean(x$samples$lib.size) * 1e-6
M <- median(x$samples$lib.size) * 1e-6
lcpm.cutoff <- log2(10/M + 2/L)

par(mfrow=c(1,2))
lcpm <- cpm(x, log=TRUE)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.40), las=2, main="", xlab="")
title(main=paste("All Genes n =", nrow(x)), xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 1:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n")
lcpm <- cpm(x_expr, log=TRUE)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.40), las=2, main="", xlab="")
title(main=paste("Expressed Genes n =", nrow(x_expr)), xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n")


# NORM FACTOR COMPUTATION AND PLOT MDS

x_norm <- calcNormFactors(x_expr, method = normalisation)

lcpm <- cpm(x_norm, log=TRUE)

col.group <- as.factor(group)
levels(col.group) <- col
col.group <- as.character(col.group)

par(mfrow=c(1,2))

if (length(grep("A", genome)) == 1){
  lcpmA = lcpm[grep(row.names(lcpm), pattern = "^A"), ]  
  boxplot(lcpmA, las=2, col=col, main="A subgenome")
}

if (length(grep("C", genome)) == 1){
  lcpmC = lcpm[grep(row.names(lcpm), pattern = "^C"), ]
  boxplot(lcpmC, las=2, col=col, main="C subgenome")
}


par(mfrow=c(1,1))
plotMDS(lcpm, labels=group, col=col)


# DEG ANALYSIS

par(mfrow=c(1,2))

design <- model.matrix(~0+group)
colnames(design) <- gsub("group", "", colnames(design))
design
contr.matrix

v <- voom(x_norm, design, plot=TRUE)

vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)

plotSA(efit)

tfit <- treat(vfit, lfc = lfc)
dt <- decideTests(tfit)

dtA = dt[grep(row.names(dt), pattern = "^A"), ]
dtC = dt[grep(row.names(dt), pattern = "^C"), ]

par(mfrow=c(1,ncol(dt)))
for (i in 1:ncol(dt)){
plotMD(tfit, column=i, status=dt[,i], main=colnames(tfit)[i], xlim=c(-8,13))
}


# OUTPUT RESULTS

out <- list()
out$tfit <- tfit
out$dt  <- dt
out$dtA <- dtA
out$dtC <- dtC
out$x_norm <- x_norm
out$cpm <- cpm(x_norm)
out$lcpm <- cpm(x_norm, log = T)
out$keep.exprs <- keep.exprs
return(out)
}


```

## DEG calculation for Parental average

```{r, DEG calculation for Parental Average}
getDEG_pmix <- function(path, files, group, columns, contr.matrix, lfc, genome, normalisation, col, keep_homeo){
  
# LOAD DATA

data_list = list()
count = 0
for (f in files){
  count = count + 1 
  file_full =  paste0(path, files[[count]])
   data_list[[f]] = suppressMessages( vroom(file_full, skip = 1, col_names = T) )
}

genes_expression =  do.call(cbind,data_list)
to_keep = seq(from = 7, to = ncol(genes_expression), by = 7)
genes_expression = genes_expression[,c(1:6,to_keep)]

colnames(genes_expression) = columns
genes_expression_basic = genes_expression
rownames(genes_expression_basic) = genes_expression_basic$gene


# FILTER GENES BY HOMEOLOGUES

if (keep_homeo == T){
homeo = vroom('/mnt/genouest/groups/bipaa/work/Acyrthosyphon_pisum/EGI/grichard/BP/STIRRER/Resource/Homeologie_Bna_DarV10_Meth.txt', col_names = F)
homeoA = as.data.frame(homeo$X1)
homeoC = as.data.frame(homeo$X2)
colnames(homeoA) = "gene"
colnames(homeoC) = "gene"
homeo = as.data.frame(rbind(homeoA, homeoC))
homeo = as.data.frame(homeo[order(homeo),])
colnames(homeo) = "gene"
homeo = unique(homeo)
rownames(homeo) = homeo$gene

genes_expression_basic = merge(genes_expression_basic, homeo, by = "row.names" )
genes_expression_basic = genes_expression_basic[,3:ncol(genes_expression_basic)-1]
colnames(genes_expression_basic)[1] = "gene"
rownames(genes_expression_basic) = genes_expression_basic$gene

}


# GENES FILTRATION DEPENDING ON CHOSEN GENOME

genes_expression_basic = genes_expression_basic[grep(genes_expression_basic$chr, pattern = genome),]

# LOAD GENES ANNOTATION FROM FEATURECOUNTS OUTPUT

genes             = genes_expression_basic[,c(1:6)]
genes[,2]         = sub(x = genes[,2],pattern = "^.*;",replacement = "")
genes[,3]         = sub(x = genes[,3],pattern = ";.*$",replacement = "")
genes[,4]         = sub(x = genes[,4],pattern = "^.*;",replacement = "")
genes[,5]         = sub(x = genes[,5],pattern = "^.*;",replacement = "")


# CREATE DGE OBJECT

counts            = genes_expression_basic[,c(7:ncol(genes_expression_basic))]
row.names(counts) <- genes_expression_basic[,1]
lib.size          = apply( genes_expression_basic[,7:ncol(genes_expression_basic)], 2, sum )
norm.factors      = rep(1, length(lib.size))
samplenames      = colnames(counts)
samples           = as.data.frame( cbind(files, group) )
rownames(samples) = samplenames

x = DGEList(counts = counts,
             lib.size = lib.size,
             genes = genes,
             samples = samples, 
             remove.zeros = F)

cat(cbind(samplenames,lib.size))

rownames(x$counts) = x$genes$gene


# KEEP ONLY EXPRESSED GENES AND PLOT

keep.exprs <- filterByExpr(x, group=group)
x_expr <- x[keep.exprs,, keep.lib.sizes=FALSE]

nsamples <- ncol(x)

L <- mean(x$samples$lib.size) * 1e-6
M <- median(x$samples$lib.size) * 1e-6
lcpm.cutoff <- log2(10/M + 2/L)

par(mfrow=c(1,2))
lcpm <- cpm(x, log=TRUE)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.40), las=2, main="", xlab="")
title(main=paste("All Genes n =", nrow(x)), xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 1:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n")
lcpm <- cpm(x_expr, log=TRUE)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.40), las=2, main="", xlab="")
title(main=paste("Expressed Genes n =", nrow(x_expr)), xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", samplenames, text.col=col, bty="n")


# NORM FACTOR COMPUTATION AND PLOT MDS

x_norm <- calcNormFactors(x_expr, method = normalisation)

lcpm <- cpm(x_norm, log=TRUE)

col.group <- as.factor(group)
levels(col.group) <- col
col.group <- as.character(col.group)

par(mfrow=c(1,2))

if (length(grep("A", genome)) == 1){
  lcpmA = lcpm[grep(row.names(lcpm), pattern = "^A"), ]  
  boxplot(lcpmA, las=2, col=col, main="A subgenome")
}

if (length(grep("C", genome)) == 1){
  lcpmC = lcpm[grep(row.names(lcpm), pattern = "^C"), ]
  boxplot(lcpmC, las=2, col=col, main="C subgenome")
}


par(mfrow=c(1,1))
plotMDS(lcpm, labels=group, col=col)

# PARENTAL MIX CALCULATION

lcpm <- cpm(x_norm, log=TRUE)

x_norm$counts = lcpm
for (r in 1:nrow(x_norm$counts)){
  x_norm$counts[r,7] = mean( x_norm$counts[r,1],x_norm$counts[r,7])
  x_norm$counts[r,8] = mean( x_norm$counts[r,2],x_norm$counts[r,8])
  x_norm$counts[r,9] = mean( x_norm$counts[r,3],x_norm$counts[r,9])
}

colnames(x_norm$counts)[7] = "Parental_mix R1"
colnames(x_norm$counts)[8] = "Parental_mix R2"
colnames(x_norm$counts)[9] = "Parental_mix R3"

x_norm$counts = x_norm$counts[, -c(1,2,3) ]

x_norm$samples[7,2] = mean(c(x_norm$samples[1,"lib.size"],x_norm$samples[7,"lib.size"]))
x_norm$samples[8,2] = mean(c(x_norm$samples[2,"lib.size"],x_norm$samples[8,"lib.size"]))
x_norm$samples[9,2] = mean(c(x_norm$samples[3,"lib.size"],x_norm$samples[9,"lib.size"]))

x_norm$samples[7,3] = mean(c(x_norm$samples[1,"norm.factors"],x_norm$samples[7,"norm.factors"]))
x_norm$samples[8,3] = mean(c(x_norm$samples[2,"norm.factors"],x_norm$samples[8,"norm.factors"]))
x_norm$samples[9,3] = mean(c(x_norm$samples[3,"norm.factors"],x_norm$samples[9,"norm.factors"]))

x_norm$samples[,1] = as.character(x_norm$samples[,1])
x_norm$samples[7,1] = "Parental_mix"
x_norm$samples[8,1] = "Parental_mix"
x_norm$samples[9,1] = "Parental_mix"
x_norm$samples[,1] = as.factor(x_norm$samples[,1])

row.names(x_norm$samples)[7] = "Parental_mix R1"
row.names(x_norm$samples)[8] = "Parental_mix R2"
row.names(x_norm$samples)[9] = "Parental_mix R3"

x_norm$samples = x_norm$samples[ -c(1, 2, 3),]

group2 = substr(colnames(x_norm$counts), 1, nchar(colnames(x_norm$counts))-3)

# DEG ANALYSIS

par(mfrow=c(1,2))

design <- model.matrix(~0+group2)
colnames(design) <- gsub("group2", "", colnames(design))

contr.matrix <- makeContrasts(
  Hybrid_vs_Parental_mix = Triploid_Ch_AAC - Parental_mix,
  levels = colnames(design)
)


v <- voom_lcpm(x_norm, design, plot=TRUE)

vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)

plotSA(efit)

tfit <- treat(vfit, lfc = lfc)
dt <- decideTests(tfit)

dtA = dt[grep(row.names(dt), pattern = "^A"), ]
dtC = dt[grep(row.names(dt), pattern = "^C"), ]

par(mfrow=c(1,ncol(dt)))
for (i in 1:ncol(dt)){
plotMD(tfit, column=i, status=dt[,i], main=colnames(tfit)[i], xlim=c(-8,13))
}


# OUTPUT RESULTS

out <- list()
out$tfit <- tfit
out$dt  <- dt
out$dtA <- dtA
out$dtC <- dtC
out$x_norm <- x_norm
out$cpm <- 2^x_norm$counts
out$lcpm <- x_norm$counts
out$keep.exprs <- keep.exprs
return(out)
}
```

## DEG Plots

```{r}

library(ggplot2)
library(ggrastr)
library(cowplot)
library(devtools)
source_gist("524eade46135f6348140")

plotDEG <- function(data, cols_1, cols_2, name_1, name_2, comp_name, genome){
  
  
  comparison = data$lcpm[,c(1:2)]
  
if(genome == "A"){
 sum = summary(data$dtA) 
 data$lcpm = data$lcpm[grep( x = row.names(data$lcpm), pattern = "^A"),]
 comparison = comparison[grep( x = row.names(comparison), pattern = "^A"),]
}

  0.
if(genome == "C"){
 sum = summary(data$dtC) 
 data$lcpm = data$lcpm[grep( x = row.names(data$lcpm), pattern = "^C"),]
 comparison = comparison[grep( x = row.names(comparison), pattern = "^C"),]
}

comparison = data$lcpm[,c(1:2)]

for (r in 1:nrow(data$lcpm)){
  comparison[r,1] = mean(data$lcpm[r,cols_1])
  comparison[r,2] = mean(data$lcpm[r,cols_2])
}

comparison = as.data.frame(comparison)
colnames(comparison) = c(name_1, name_2)

if (genome == "A"){
comparison[grep( x = row.names(data$lcpm), pattern = "^A"),"genome"] =  paste("A -", sum[3,comp_name], "Up &", sum[1,comp_name],"Down DEGs")
}

if (genome == "C"){
comparison[grep( x = row.names(data$lcpm), pattern = "^C"),"genome"] =  paste("C -", sum[3,comp_name], "Up &", sum[1,comp_name],"Down DEGs")
}

comparison["differential"] = "No diff. expression"
comparison[which(data$dt[,comp_name] ==  1),"differential"] = "DEG Up"
comparison[which(data$dt[,comp_name] == -1),"differential"] = "DEG Down"

comparison = na.omit(comparison)

ggplot(comparison, aes(x = comparison[,2], y = comparison[,1]))+
  facet_wrap(genome ~ .) +
  geom_point_rast(aes( color = differential, alpha = differential)) +
  theme_cowplot() +
  scale_color_manual(paste0("Genes, n=",nrow(comparison)), values = c("blue","red","black")) +
  scale_alpha_manual(paste0("Genes, n=",nrow(comparison)), values = c(1,1,0.1)) +
  xlab(paste0(name_2, " log2(cpm)")) +
  ylab(paste0(name_1, " log2(cpm)"))+
  ggtitle(paste0(name_1, " vs ", name_2)) +
  stat_smooth_func(geom="text",method="lm",hjust=0,parse=TRUE,ypos = max(comparison[,1:2])+0.5, xpos = min(comparison[,1:2])) +
  geom_smooth(method="lm",se=FALSE, color = "grey") +
  theme(plot.title = element_text(size = 12))
}

```

# Comparisons for the A subgenome

## Gene by Gene Level

We must filter out genes to only keep expressed genes and homeologs to have a consistent list of genes between the gene by gene and homeoglobal scales. We must first find the least biased way to filter genes and to normalize their expression.


### How to normalise expression on the A subgenome


#### All comparison only A subgenome from start

```{r, fig.height= 5, fig.width= 8}

path =  "/mnt/genouest/groups/bipaa/work/Acyrthosyphon_pisum/EGI/grichard/BP/STIRRER/RNA-seq/SnakePipes/All_samples_ete_1M/featureCounts/"

files = list.files(path = path, pattern = "CS.*.counts.txt$|DhB.*.counts.txt$|DCh.*.counts.txt$|DCD.*.counts.txt$|CD4.*.counts.txt$")

col <- c( brewer.pal(8, "Blues")[6:8], brewer.pal(8, "Greens")[6:8], brewer.pal(8, "Reds")[6:8], brewer.pal(8, "Oranges")[6:8], brewer.pal(8, "Greys")[6:8] )

group   = c(rep("Darmor_AA",3),rep("Chiifu_AA",3),rep("Triploid_D_AAC",3),rep("Triploid_Ch_AAC",3),rep("Darmor_AACC",3))

columns = c("gene","chr","start","end","strand","length",
            "Darmor_AA R1", "Darmor_AA R2", "Darmor_AA R3", 
            "Chiifu_AA R1","Chiifu_AA R2","Chiifu_AA R3",
            "Triploid_D_AAC R1", "Triploid_D_AAC R2", "Triploid_D_AAC R3", 
            "Triploid_Ch_AAC R1", "Triploid_Ch_AAC R2", "Triploid_Ch_AAC R3", 
            "Darmor_AACC R1","Darmor_AACC R2","Darmor_AACC R3")

design <- model.matrix(~0+group)
colnames(design) <- gsub("group", "", colnames(design))

contr.matrix <- makeContrasts(
  Chiifu_vs_Darmor = Chiifu_AA - Darmor_AACC,
  TriploidCh_vs_Darmor = Triploid_Ch_AAC -  Darmor_AACC,
  TriploidCh_vs_Chiifu = Triploid_Ch_AAC - Chiifu_AA,
  Darmor_vs_Darmor_AA = Darmor_AACC - Darmor_AA,
  Darmor_vs_TriploidD = Darmor_AACC -  Triploid_D_AAC,
  TriploidCh_vs_TriploidD = Triploid_Ch_AAC - Triploid_D_AAC,
  Chiifu_vs_Darmor_AA = Chiifu_AA - Darmor_AA,
  levels = colnames(design)
)

lfc=0.584962501

DEG_all_A <- getDEG(path, files, group, columns, contr.matrix, lfc, genome = "^A", normalisation = "RLE", col, keep_homeo= F)


kable_minimal(kable(summary(DEG_all_A$dtA),caption = "A subgenome"))


to_plot <- as.data.frame(DEG_all_A$lcpm)

to_plot = gather(as.data.frame(to_plot), sample, lcpm, "Darmor_AA R1":"Darmor_AACC R3",factor_key=TRUE)

suppressMessages( ggplot(to_plot, aes(x = to_plot$sample, y = to_plot$lcpm, color = to_plot$sample)) +
  geom_boxplot(width = 0.60, alpha = 0.3)+
  geom_violin( width = 0.95, fill = NA)+
  scale_color_manual(values = col) +
  theme_cowplot()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none")+
  xlab("Samples") +
  ylab("Log2 CPM") +
  ggtitle("Expression on Darmor-bzh A subgenome") )

suppressMessages( ggplot(to_plot, aes( x = to_plot$lcpm, color = to_plot$sample)) +
  geom_density(alpha =0.2)+
  scale_color_manual("Samples", values = col) +
  scale_fill_manual("Samples", values = col) +
  xlim(min(to_plot$lcpm)-1,max(to_plot$lcpm))+
  theme_cowplot()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("samples") +
  ylab("Log2 CPM") +
  ggtitle("Expression on Darmor-bzh A subgenome") )



```

Here we can see that Chiifu is biased, especially on the box plots. Indeed there are many genes that are not expressed. Let's see how to fix this.

#### All comparison A/C subgenome and looking at A subsequently

In that section we normalize the expression of all samples on both A and C subgenomes and only keep A genes afterwards.

```{r, fig.height= 5, fig.width= 8}

path =  "/mnt/genouest/groups/bipaa/work/Acyrthosyphon_pisum/EGI/grichard/BP/STIRRER/RNA-seq/SnakePipes/All_samples_ete_1M/featureCounts/"

files = list.files(path = path, pattern = "CS.*.counts.txt$|DhB.*.counts.txt$|DCh.*.counts.txt$|DCD.*.counts.txt$|CD4.*.counts.txt$")

col <- c( brewer.pal(8, "Blues")[6:8], brewer.pal(8, "Greens")[6:8], brewer.pal(8, "Reds")[6:8], brewer.pal(8, "Oranges")[6:8], brewer.pal(8, "Greys")[6:8] )

group   = c(rep("Darmor_AA",3),rep("Chiifu_AA",3),rep("Triploid_D_AAC",3),rep("Triploid_Ch_AAC",3),rep("Darmor_AACC",3))

columns = c("gene","chr","start","end","strand","length",
            "Darmor_AA R1", "Darmor_AA R2", "Darmor_AA R3", 
            "Chiifu_AA R1","Chiifu_AA R2","Chiifu_AA R3",
            "Triploid_D_AAC R1", "Triploid_D_AAC R2", "Triploid_D_AAC R3", 
            "Triploid_Ch_AAC R1", "Triploid_Ch_AAC R2", "Triploid_Ch_AAC R3", 
            "Darmor_AACC R1","Darmor_AACC R2","Darmor_AACC R3")

design <- model.matrix(~0+group)
colnames(design) <- gsub("group", "", colnames(design))

contr.matrix <- makeContrasts(
  Chiifu_vs_Darmor = Chiifu_AA - Darmor_AACC,
  TriploidCh_vs_Darmor = Triploid_Ch_AAC -  Darmor_AACC,
  TriploidCh_vs_Chiifu = Triploid_Ch_AAC - Chiifu_AA,
  Darmor_vs_Darmor_AA = Darmor_AACC - Darmor_AA,
  Darmor_vs_TriploidD = Darmor_AACC -  Triploid_D_AAC,
  TriploidCh_vs_TriploidD = Triploid_Ch_AAC - Triploid_D_AAC,
  Chiifu_vs_Darmor_AA = Chiifu_AA - Darmor_AA,
  levels = colnames(design)
)

lfc=0.584962501

DEG_all_A <- getDEG(path, files, group, columns, contr.matrix, lfc, genome = ".*", normalisation = "RLE", col, keep_homeo = F)

kable_minimal(kable(summary(DEG_all_A$dtA),caption = "A subgenome"))


to_plot <- as.data.frame(DEG_all_A$lcpm)
to_plot = to_plot[grep(row.names(to_plot), pattern = "^A"), ]  

to_plot = gather(as.data.frame(to_plot), sample, lcpm, "Darmor_AA R1":"Darmor_AACC R3",factor_key=TRUE)

ggplot(to_plot, aes(x = to_plot$sample, y = to_plot$lcpm, color = to_plot$sample)) +
  geom_boxplot(width = 0.60, alpha = 0.3)+
  geom_violin( width = 0.95, fill = NA)+
  scale_color_manual(values = col) +
  theme_cowplot()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none")+
  xlab("Samples") +
  ylab("Log2 CPM") +
  ggtitle("Expression on Darmor-bzh A subgenome")

ggplot(to_plot, aes( x = to_plot$lcpm, color = to_plot$sample)) +
  geom_density(alpha =0.2)+
  scale_color_manual("Samples", values = col) +
  scale_fill_manual("Samples", values = col) +
  xlim(min(to_plot$lcpm)-1,max(to_plot$lcpm))+
  theme_cowplot()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("samples") +
  ylab("Log2 CPM") +
  ggtitle("Expression on Darmor-bzh A subgenome")



```

We can see a slight improvement on Chiifu expression vs others.


#### Removing uncovered regions in common between Chiifu

Now we try a more stringent approach. We only keep the regions covered in Chiifu and Darmor-bzh. The files and script to generate them is available here (get_covered_regions.Rmd): /home/genouest/inra_umr1349/grichard/Data/BP/STIRRER/RNA-seq/SnakePipes/Analyses/regions_without_coverage_in_chiifu/coverage_bedgraphs

```{r}
path =  "/mnt/genouest/groups/bipaa/work/Acyrthosyphon_pisum/EGI/grichard/BP/STIRRER/RNA-seq/SnakePipes/All_samples_ete_1M/featureCounts_on_covered_regions/"
files = list.files(path = path, pattern = "CS.*chiifu.txt$|DhB.*chiifu.txt$|DCh.*chiifu.txt$|DCD.*chiifu.txt$|CD4.*chiifu.txt$")

col <- c( brewer.pal(8, "Blues")[6:8], brewer.pal(8, "Greens")[6:8], brewer.pal(8, "Reds")[6:8], brewer.pal(8, "Oranges")[6:8], brewer.pal(8, "Greys")[6:8] )

group   = c(rep("Darmor_AA",3),rep("Chiifu_AA",3),rep("Triploid_D_AAC",3),rep("Triploid_Ch_AAC",3),rep("Darmor_AACC",3))

columns = c("gene","chr","start","end","strand","length",
            "Darmor_AA R1", "Darmor_AA R2", "Darmor_AA R3", 
            "Chiifu_AA R1","Chiifu_AA R2","Chiifu_AA R3",
            "Triploid_D_AAC R1", "Triploid_D_AAC R2", "Triploid_D_AAC R3", 
            "Triploid_Ch_AAC R1", "Triploid_Ch_AAC R2", "Triploid_Ch_AAC R3", 
            "Darmor_AACC R1","Darmor_AACC R2","Darmor_AACC R3")

design <- model.matrix(~0+group)
colnames(design) <- gsub("group", "", colnames(design))

contr.matrix <- makeContrasts(
  Chiifu_vs_Darmor = Chiifu_AA - Darmor_AACC,
  TriploidCh_vs_Darmor = Triploid_Ch_AAC -  Darmor_AACC,
  TriploidCh_vs_Chiifu = Triploid_Ch_AAC - Chiifu_AA,
  Darmor_vs_Darmor_AA = Darmor_AACC - Darmor_AA,
  Darmor_vs_TriploidD = Darmor_AACC -  Triploid_D_AAC,
  TriploidCh_vs_TriploidD = Triploid_Ch_AAC - Triploid_D_AAC,
  Chiifu_vs_Darmor_AA = Chiifu_AA - Darmor_AA,
  levels = colnames(design)
)

lfc=0.584962501

DEG_all_A_on_chiifu <- getDEG(path, files, group, columns, contr.matrix, lfc, genome = ".*", normalisation = "RLE", col, keep_homeo = F)

kable_minimal(kable(summary(DEG_all_A_on_chiifu$dtA),caption = "A subgenome"))


to_plot <- as.data.frame(DEG_all_A_on_chiifu$lcpm)
to_plot = to_plot[grep(row.names(to_plot), pattern = "^A"), ]  

to_plot = gather(as.data.frame(to_plot), sample, lcpm, "Darmor_AA R1":"Darmor_AACC R3",factor_key=TRUE)

ggplot(to_plot, aes(x = to_plot$sample, y = to_plot$lcpm, color = to_plot$sample)) +
  geom_boxplot(width = 0.60, alpha = 0.3)+
  geom_violin( width = 0.95, fill = NA)+
  scale_color_manual(values = col) +
  theme_cowplot()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none")+
  xlab("Samples") +
  ylab("Log2 CPM") +
  ggtitle("Expression on Darmor-bzh A subgenome w/ coverage")

ggplot(to_plot, aes( x = to_plot$lcpm, color = to_plot$sample)) +
  geom_density(alpha =0.2)+
  scale_color_manual("Samples", values = col) +
  scale_fill_manual("Samples", values = col) +
  theme_cowplot()+
  xlim(min(to_plot$lcpm)-1,max(to_plot$lcpm))+
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("samples") +
  ylab("Log2 CPM") +
  ggtitle("Expression on Darmor-bzh A subgenome w/ coverage")

```

Using regions covered in chiifu we have a better balance of expression between chiifu and the other genotypes, thus helping to mitigate the bias of mapping Chiifu on the Darmor-bzh genome. We keep 17348 genes overall on the A subgenome.

### Keeping only homeologs

We need to test if it's better to filter homeologs before or after normalization of genes expression.

#### Filtration before normalization

```{r}

DEG_all_A_on_chiifu_homeo <- getDEG(path, files, group, columns, contr.matrix, lfc, genome = ".*", normalisation = "RLE", col, keep_homeo = T)

kable_minimal(kable(summary(DEG_all_A_on_chiifu_homeo$dtA),caption = "A subgenome"))


to_plot <- as.data.frame(DEG_all_A_on_chiifu_homeo$lcpm)
to_plot = to_plot[grep(row.names(to_plot), pattern = "^A"), ]  

to_plot = gather(as.data.frame(to_plot), sample, lcpm, "Darmor_AA R1":"Darmor_AACC R3",factor_key=TRUE)

ggplot(to_plot, aes(x = to_plot$sample, y = to_plot$lcpm, color = to_plot$sample)) +
  geom_boxplot(width = 0.60, alpha = 0.3)+
  geom_violin( width = 0.95, fill = NA)+
  scale_color_manual(values = col) +
  theme_cowplot()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none")+
  xlab("Samples") +
  ylab("Log2 CPM") +
  ggtitle("Expression on Darmor-bzh A subgenome w/ coverage")

ggplot(to_plot, aes( x = to_plot$lcpm, color = to_plot$sample)) +
  geom_density(alpha =0.2)+
  scale_color_manual("Samples", values = col) +
  scale_fill_manual("Samples", values = col) +
  theme_cowplot()+
  xlim(min(to_plot$lcpm)-1,max(to_plot$lcpm))+
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("samples") +
  ylab("Log2 CPM") +
  ggtitle("Expression on Darmor-bzh A subgenome w/ coverage")


```


#### Filtration after normalization

```{r}
homeo = vroom('/mnt/genouest/groups/bipaa/work/Acyrthosyphon_pisum/EGI/grichard/BP/STIRRER/Resource/Homeologie_Bna_DarV10_Meth.txt', col_names = F)
homeoA = as.data.frame(homeo$X1)
homeoC = as.data.frame(homeo$X2)
colnames(homeoA) = "gene"
colnames(homeoC) = "gene"
homeo = as.data.frame(rbind(homeoA, homeoC))
homeo = as.data.frame(homeo[order(homeo),])
colnames(homeo) = "gene"
homeo = unique(homeo)
rownames(homeo) = homeo$gene

DEG_homeo_A_lpcm = merge(DEG_all_A_on_chiifu$lcpm, homeo, by = "row.names" )
rownames(DEG_homeo_A_lpcm) = DEG_homeo_A_lpcm$gene
DEG_homeo_A_lpcm = DEG_homeo_A_lpcm[,3:ncol(DEG_homeo_A_lpcm)-1]
DEG_homeo_A_lpcm = as.data.frame(DEG_homeo_A_lpcm)

DEG_homeo_A_dtA = merge(DEG_all_A_on_chiifu$dtA, homeo, by = "row.names" )
rownames(DEG_homeo_A_dtA ) = DEG_homeo_A_dtA $gene
DEG_homeo_A_dtA  = DEG_homeo_A_dtA [,3:ncol(DEG_homeo_A_dtA )-1]

DEG_homeo_A_dt = merge(DEG_all_A_on_chiifu$dt, homeo, by = "row.names" )
rownames(DEG_homeo_A_dt ) = DEG_homeo_A_dt $gene
DEG_homeo_A_dt  = DEG_homeo_A_dt [,3:ncol(DEG_homeo_A_dt )-1]



to_plot <- as.data.frame(DEG_homeo_A_lpcm)
to_plot = to_plot[grep(row.names(to_plot), pattern = "^A"), ]  

to_plot = gather(as.data.frame(to_plot), sample, lcpm, "Darmor_AA R1":"Darmor_AACC R3",factor_key=TRUE)


ggplot(to_plot, aes(x = to_plot$sample, y = to_plot$lcpm, color = to_plot$sample)) +
  geom_boxplot(width = 0.60, alpha = 0.3)+
  geom_violin( width = 0.95, fill = NA)+
  scale_color_manual(values = col) +
  theme_cowplot()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none")+
  xlab("Samples") +
  ylab("Log2 CPM") +
  ggtitle("Expression on Darmor-bzh A subgenome w/ coverage")
```

No real difference, after normalization is way more complicated to implement, so we keep the filtration before normalization since changes will only be marginal anyways

#### DEG plots A subgenome using regions covered in Chiifu

```{r, fig.height= 4, fig.width= 6}

data = DEG_all_A_on_chiifu_homeo
cols_1 = c("Triploid_Ch_AAC R1", "Triploid_Ch_AAC R2", "Triploid_Ch_AAC R3")
cols_2 = c("Darmor_AACC R1","Darmor_AACC R2", "Darmor_AACC R3")
name_1 = "Hybrid DCh AAC"
name_2 = "Darmor-bzh AACC"
genome = "A"
comp_name = "TriploidCh_vs_Darmor"
p1 <- plotDEG(data= data, cols_1= cols_1 , cols_2= cols_2, name_1 = name_1, name_2 = name_2, genome = "A", comp_name = comp_name)
p1
cols_1 = c("Chiifu_AA R1", "Chiifu_AA R2", "Chiifu_AA R3")
cols_2 = c("Darmor_AACC R1","Darmor_AACC R2", "Darmor_AACC R3")
name_1 = "Chiifu AA"
name_2 = "Darmor-bzh AACC"
genome = "A"
comp_name = "Chiifu_vs_Darmor"
p2 <- plotDEG(data= data, cols_1= cols_1 , cols_2= cols_2, name_1 = name_1, name_2 = name_2, genome = "A", comp_name = comp_name)
p2
cols_2 = c("Chiifu_AA R1", "Chiifu_AA R2", "Chiifu_AA R3")
cols_1 = c("Triploid_Ch_AAC R1","Triploid_Ch_AAC R2", "Triploid_Ch_AAC R3")
name_2 = "Chiifu AA"
name_1 = "Hybrid DCh AAC"
genome = "A"
comp_name = "TriploidCh_vs_Chiifu"
p3 <- plotDEG(data= data, cols_1= cols_1 , cols_2= cols_2, name_1 = name_1, name_2 = name_2, genome = "A", comp_name = comp_name)
p3




```


## Parental mix raw genes coverage calculation

To compare Triploid vs Tetra and diploid genomes, it is necessary to compute the parental mix genes expression. Replicates number (1, 2, 3) correspond to different batches of plants that were raised in the same conditions among the three genotypes. For each gene, we thus calculate the parental mix expression by computing the average of the Counts Per Million (CPM), per replicate, between the tetraploid and the diploid genotypes.

The genes expression were obtained using the SnakePipes mRNA-seq pipeline on the Darmor-bzh v10 genome assembly, using STAR with the following parameters for the mapping: `--alignEndsType EndToEnd --outFilterMismatchNmax 1`.

```{r}
path =  "/mnt/genouest/groups/bipaa/work/Acyrthosyphon_pisum/EGI/grichard/BP/STIRRER/RNA-seq/SnakePipes/All_samples_ete_1M/featureCounts_on_covered_regions/"
files = list.files(path = path, pattern = "CS.*chiifu.txt$|DhB.*chiifu.txt$|DCh.*chiifu.txt$")

col <- c( brewer.pal(8, "Greens")[6:8], brewer.pal(8, "Reds")[6:8], brewer.pal(8, "Greys")[6:8] )

group   = c(rep("Chiifu_AA",3),rep("Triploid_Ch_AAC",3),rep("Darmor_AACC",3))

columns = c("gene","chr","start","end","strand","length",
            "Chiifu_AA R1","Chiifu_AA R2","Chiifu_AA R3",
            "Triploid_Ch_AAC R1", "Triploid_Ch_AAC R2", "Triploid_Ch_AAC R3", 
            "Darmor_AACC R1","Darmor_AACC R2","Darmor_AACC R3")

design <- model.matrix(~0+group)
colnames(design) <- gsub("group", "", colnames(design))

lfc=0.584962501

DEG_parental_homeo <- getDEG_pmix(path, files, group, columns, contr.matrix, lfc, genome = ".*", normalisation = "RLE", col, keep_homeo = T)

kable_minimal(kable(summary(DEG_parental_homeo$dtA),caption = "A subgenome"))


to_plot <- as.data.frame(DEG_parental_homeo$lcpm)
to_plot = to_plot[grep(row.names(to_plot), pattern = "^A"), ]  

to_plot = gather(as.data.frame(to_plot), sample, lcpm, "Triploid_Ch_AAC R1":"Parental_mix R3",factor_key=TRUE)

ggplot(to_plot, aes(x = to_plot$sample, y = to_plot$lcpm, color = to_plot$sample)) +
  geom_boxplot(width = 0.60, alpha = 0.3)+
  geom_violin( width = 0.95, fill = NA)+
  scale_color_manual(values = col) +
  theme_cowplot()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none")+
  xlab("Samples") +
  ylab("Log2 CPM") +
  ggtitle("Expression on Darmor-bzh A subgenome w/ coverage")

ggplot(to_plot, aes( x = to_plot$lcpm, color = to_plot$sample)) +
  geom_density(alpha =0.2)+
  scale_color_manual("Samples", values = col) +
  scale_fill_manual("Samples", values = col) +
  theme_cowplot()+
  xlim(min(to_plot$lcpm)-1,max(to_plot$lcpm))+
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("samples") +
  ylab("Log2 CPM") +
  ggtitle("Expression on Darmor-bzh A subgenome w/ coverage")

```


```{r, fig.height=4, fig.width=6}
data = DEG_parental_homeo
cols_1 = c("Triploid_Ch_AAC R1", "Triploid_Ch_AAC R2", "Triploid_Ch_AAC R3")
cols_2 = c("Parental_mix R1","Parental_mix R2", "Parental_mix R3")
name_1 = "Hybrid DCh AAC"
name_2 = "Parental mix"
genome = "A"
comp_name = "Hybrid_vs_Parental_mix"
p1 <- plotDEG(data= data, cols_1= cols_1 , cols_2= cols_2, name_1 = name_1, name_2 = name_2, genome = "A", comp_name = comp_name)
p1

```


## Parental dominance classes


Additivité :

a) Catégorie no change : si pas de différence signif entre Darmor et Chiifu + pas de différence signif entre (Ch + Darmor)/2 et Dch
b) Catégorie I: Darmor est uprégulé par rapport à chiifu + pas de différence signif entre (Ch + Darmor)/2 et Dch
c) Catégorie XII : Darmor est downrégulé par rapport à chiifu + pas de différence signif entre Darmor et Dch​

Dominance maternelle (cf Darmor) :

a) Catégorie II: Darmor est uprégulé par rapport à chiifu + différence signif entre (Ch + Darmor)/2 et Dch + pas de différence signif entre Darmor et Dch
b) Catégorie XI: Darmor est down-régulé par rapport à chiifu + différence signif entre (Ch + Darmor)/2 et Dch​ + pas de différence signif entre Darmor et Dch

Dominance paternelle (cf Chiifu) :

a) Catégorie IV: Darmor est down-régulé par rapport à chiifu + différence signif entre (Ch + Darmor)/2 et Dch + pas de différence signif entre Chiifu et Dch
b) Catégorie IX: Darmor est up-régulé par rapport à chiifu + différence signif entre (Ch + Darmor)/2 et Dch​ + pas de différence signif entre Chiifu et Dch

Transgressivité down : 

a) Catégorie III: Darmor est uprégulé par rapport à chiifu + différence signif entre (Ch + Darmor)/2 et Dch + Dch est downrégulé par rapport à Darmor + Dch est downrégulé par rapport à Chiifu
b) Catégorie VII: pas de diff signif entre Darmor et Chiifu + différence signif entre (Ch + Darmor)/2 et Dch + Dch est downrégulé par rapport à Darmor + Dch est downrégulé par rapport à Chiifu
c) Catégorie X: Darmor est downrégulé par rapport à chiifu + différence signif entre (Ch + Darmor)/2 et Dch + Dch est downrégulé par rapport à Darmor + Dch est downrégulé par rapport à Chiifu

Transgressivité up : 

a) Catégorie V: Darmor est uprégulé par rapport à chiifu + différence signif entre (Ch + Darmor)/2 et Dch + Dch est uprégulé par rapport à Darmor + Dch est uprégulé par rapport à Chiifu
b) Catégorie VI: Darmor est downrégulé par rapport à chiifu + différence signif entre (Ch + Darmor)/2 et Dch + Dch est uprégulé par rapport à Darmor + Dch est uprégulé par rapport à Chiifu
c) Catégorie VIII: pas de diff signif entre Darmor et Chiifu + différence signif entre (Ch + Darmor)/2 et Dch + Dch est uprégulé par rapport à Darmor + Dch est uprégulé par rapport à Chiifu



```{r}

getDEGClasses <- function(d){

d = as.data.frame(d)

d["Class"] = NA

d[which(
  d[,"Hy/P"] == 0 &
  d[,"Ch/Da"] == 0 &
  d[,"Hy/Da"] == 0 &
  d[,"Hy/Ch"] == 0  
),"Class"] = "No Change"

d[which(
  d[,"Hy/P"] == 0 &
  d[,"Ch/Da"] == -1
),"Class"] = "I"

d[which(
  d[,"Hy/P"] == 0 &
  d[,"Ch/Da"] == 1
),"Class"] = "XII"

d[which(
  d[,"Hy/Da"] == 0 &
  d[,"Ch/Da"] == -1
),"Class"] = "II"

d[which(
  d[,"Hy/Da"] == 0 &
  d[,"Ch/Da"] == 1
),"Class"] = "XI"

d[which(
  d[,"Hy/Ch"] == 0 &
  d[,"Ch/Da"] == 1
),"Class"] = "IV"

d[which(
  d[,"Hy/Ch"] == 0 &
  d[,"Ch/Da"] == -1
),"Class"] = "IX"

d[which(
  d[,"Hy/P"] == -1 & 
  d[,"Hy/Ch"] == -1 &  
  d[,"Hy/Da"] == -1 &
  d[,"Ch/Da"] == -1
),"Class"] = "III"

d[which(
  d[,"Hy/P"] == -1 & 
  d[,"Hy/Ch"] == -1 &  
  d[,"Hy/Da"] == -1 &
  d[,"Ch/Da"] == 0
),"Class"] = "VII"

d[which(
  d[,"Hy/P"] == -1 & 
  d[,"Hy/Ch"] == -1 &  
  d[,"Hy/Da"] == -1 &
  d[,"Ch/Da"] == 1
),"Class"] = "X"

d[which(
  d[,"Hy/P"] == 1 &
  d[,"Hy/Ch"] == 1 &  
  d[,"Hy/Da"] == 1 &
  d[,"Ch/Da"] == -1
),"Class"] = "V"

d[which(
  d[,"Hy/P"] == 1 & 
  d[,"Hy/Ch"] == 1 &  
  d[,"Hy/Da"] == 1 &
  d[,"Ch/Da"] == 1
),"Class"] = "VI"

d[which(
  d[,"Hy/P"] == 1 &
  d[,"Hy/Ch"] == 1 &  
  d[,"Hy/Da"] == 1 &
  d[,"Ch/Da"] == 0
),"Class"] = "VIII"

d[which(is.na(d[,"Class"])),"Class"] = "No expression level dominance"

d[,"combination"] = paste(d[,1],d[,2],d[,3],d[,4],d[,5], sep = "_")

return(d)
}

d = merge(DEG_parental_homeo$dtA, DEG_all_A_on_chiifu_homeo$dtA[,c(1,2,3)], by = "row.names")
colnames(d) = c("genes", "Hy/P", "Ch/Da", "Hy/Da", "Hy/Ch")
d = getDEGClasses(d)


cat("\n")
summary(as.factor(d[,"Class"]))

summary(d[,])

```

522 genes (3.2%) fell into an expression level dominance category. Out of them, a majority of transgressive up-regulated genes (Classes V, VI and VIII) and transgressive-down regulated genes (Classes III, VII and X) are found. We didn't find any Additivity, or Parental expression level dominance. The most dominant categories correspond to genes upregulated in the Hybrid vs its parents, and that are either upregulated in Darmor-bzh compared to Chiifu or that are not significantly differentially expressed between the two parents. The other two dominant categories are the opposite.

Genes that are upregulated in Darmor-bzh compared to Chiifu and the Hybrid but that are not different between the Hybrid and Chiifu are the most common (2137).
